Fix inaccurate stack-popping in custom element upgrades

domenic opened the pull request:
Fixes #1238. Usually, the popped stack entry is not the element, but
instead an already-constructed marker. This makes the stack-popping
generic, and also adds a note explaining exactly what could be going on
here.

@dominiccooney to review. FYI @rniwa since you came up with this algorithm and I messed it up when putting it into the spec.

dominiccooney commented:
This LGTM, although I have a doubts about this part of the spec.

For example, the note is right but doesn't mention that the item being popped may be an element if the author correctly calls super, but then [HTMLElement constructor](https://html.spec.whatwg.org/multipage/dom.html#htmlelement) step 7 throws an exception doing *element*.[\[SetPrototypeOf]]\(*prototype*). But I am unsure when that could happen in practice, given step 2. (And it seems like calling super multiple times is OK in that case? How would an author write that?)

domenic commented:
Oh, wow, so you're right that it's impossible to author a class that calls super() multiple times. The second super() call will error before ever calling the constructor.

However, it's still possible to trigger that error case. I've added an example to the spec of how to do it.

In general I want to make these crazy edge cases as clear as possible for implementers in the spec. Would love your thoughts on the new patch which tries to do that. (I added the fact that exceptions thrown in HTMLElement() will impact the stack, as you said.)

I wonder if we should throw in "upgrade an element" step 4, if the last entry in the stack is not an already-constructed marker? It seems like the only way that could be possible is if something horrible went wrong, and failing might be better.

dominiccooney commented on source:
```diff
@@ -66430,8 +66449,24 @@ fetch(articleURL)
+   </li>
 
    <li><p>If <var>constructResult</var> is an abrupt completion, then return
    <var>constructResult</var> (i.e., rethrow the exception).</p></li>
```
This LGTM

dominiccooney commented:
OK, I looked at pull request 1242 and it looks good.

domenic commented:
It turns out you can call `super()` twice and it will execute HTMLElement twice. It will just throw an error after the second execution :-S. So I need to fix this up a bit more.

But thanks for the review @dominiccooney; you're off the hook on this PR :)

domenic commented:
OK, this is ready for editor review; I've added another example explaining how you can call super() multiple times.

domenic commented:
@dominiccooney did you have an opinion on

> I wonder if we should throw in "upgrade an element" step 4, if the last entry in the stack is not an already-constructed marker? It seems like the only way that could be possible is if something horrible went wrong, and failing might be better.

domenic commented:
In particular, this would mean that cases like

```html
<!DOCTYPE html>
<x-tag></x-tag>

<script>
customElements.define("x-tag", class XTag extends HTMLElement {
    constructor() {
        super();
        try {
            super();
        } catch (e) {
            // purposefully swallow exception so it doesn't reach "upgrade an element" step 5
        }
    }
});
</script>
```

would terminate the "upgrade an element" algorithm after step 4, instead of moving on and setting the state to `"custom"` and calling the custom element callbacks.

annevk commented on source:
```diff
@@ -66388,8 +66424,24 @@ fetch(articleURL)
+
+    <div class="note">
+     <p>Assuming <var>C</var> calls <code data-x="">super()</code> (as it <a
+     href="#custom-element-conformance">should</a>), and that the call succeeds, this will be the
```
s/should/has to/

domenic replied:
I guess it should be "must", since this is a normative authoring conformance requirement? "has to" is not really correct since authors don't have to do what we tell them. I guess "must" is a bit dubious from that perspective too.

Maybe "will if it is conformant".

domenic replied:
pushed a fixup

annevk replied:
The main reason I said this is because you can't require things in notes.

domenic replied:
Sure, it's just referring to a requirement defined elsewhere though.

annevk commented:
We are still waiting for @dominiccooney to review and answer the question you raised, no?

domenic commented:
He reviewed, and I think we can do any further change as a separate PR. This is more if a bug fix/clarification.
